cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/SubstationIdMapping.java

```diff
@@ -10,70 +10,302 @@
+        String substationId2 = context.cgmes().substation(t2, context.nodeBreaker());
+
+        // Null could be received as substation at the boundary
+        if (substationId1 == null || substationId2 == null || substationId1.equals(substationId2)) {
+            return;
+        }
+        addAdjacency(substationAdjacency, substationId1, substationId2);
+    }
+
+    private String getVoltageLevelFromNode(String node, CgmesTerminal t) {
+        String voltageLevelId = null;
+        if (node != null && !context.boundary().containsNode(node)) {
+            voltageLevelId = context.cgmes().voltageLevel(t, context.nodeBreaker());
+        }
+        return voltageLevelId;
+    }
+
+    // Two different substations are adjacent if they are connected by a transformer
+    private void addEnds(Map<String, List<String>> substationAdjacency, PropertyBags tends) {
+        List<String> substationsIds = substationsIds(tends);
+        if (substationsIds.size() <= 1) {
+            return;
+        }
+        String sub0 = substationsIds.get(0);
+        for (int i = 1; i < substationsIds.size(); i++) {
+            String subi = substationsIds.get(i);
+
+            if (sub0.contentEquals(subi)) {
+                continue;
+            }
+            addAdjacency(substationAdjacency, sub0, subi);
+        }
+    }
+
+    // Record in the adjacency Map that "id1 is adjacent to id2" and "id2 is adjacent to id1"
+    private static void addAdjacency(Map<String, List<String>> adjacency, String id1, String id2) {
+        List<String> ad1 = adjacency.get(id1);
+        if (ad1 == null) {
+            throw new PowsyblException("Unexpected reference to Substation or voltageLevel " + id1
+                + ". It has not been defined in CGMES substations / voltageLevels.");
+        }
+        List<String> ad2 = adjacency.get(id2);
+        if (ad2 == null) {
+            throw new PowsyblException("Unexpected reference to Substation or voltageLevel " + id2
+                + ". It has not been defined in CGMES substations / voltageLevels.");
+        }
+
+        ad1.add(id2);
+        ad2.add(id1);
+    }
+
+    private void buildVoltageLevel(Map<String, List<String>> voltageLevelAdjacency) {
+        Set<String> visitedVoltageLevels = new HashSet<>();
+        voltageLevelAdjacency.keySet().forEach(vl -> {
+            if (!visitedVoltageLevels.contains(vl)) {
+                ArrayList<String> vlAds = allConnected(voltageLevelAdjacency, visitedVoltageLevels, vl);
+                String selectedVoltageLevelId = representativeVoltageLevelId(vlAds);
+                for (String voltageLevelId : vlAds) {
+                    if (!voltageLevelId.equals(selectedVoltageLevelId)) {
+                        voltageLevelMapping.put(voltageLevelId, selectedVoltageLevelId);
                     }
-                });
-        if (!mapping.isEmpty()) {
+                }
+            }
+        });
```

Comment:
Using lambda is not necessary here. Replace the forEach by a classic for-loop to increase readability, or use functional code as much as you can.

Reply:
Done